src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -120,7 +120,7 @@ public final void reset(Collection<? extends Whitelist> delegates) {
                     ew.clearCache();
                 } else if (delegate instanceof ProxyWhitelist) {
                     ProxyWhitelist pw = (ProxyWhitelist) delegate;
-                    pw.wrappers.put(this, null);
+                    pw.addWrapper(this, null);
```

Comment:
I think 124 through 130  need to be covered by read lock  on  `delegate.lock`.  
We should take a write lock in delegate here and then downgrade to a read lock for the following lines.